### PR TITLE
Homework-12. Development and testing ansible role and playbook.

### PR DESCRIPTION
**Pull request by ivanov2103**
# Выполнено ДЗ № 12

- [*] Основное ДЗ
- [*] Задание со *

## В процессе сделано:

Local infrastructure was deployed with use Vagrant.
Phyton installing moved to dedicated playbook, added tasks for installing and configuration Mongo DB, installing Ruby and configuration Puma.
Added parameter for deploy user and changed templates and tasks for use it as variable.
I use WSL and needed in some additional tuning: added environment variable (https://www.vagrantup.com/docs/other/wsl.html) and changed Vagrantfile - turning off the serial port (https://github.com/joelhandwell/ubuntu_vagrant_boxes/issues/1).
*
Wariable nginx_sites format was difined by example from https://www.vagrantup.com/docs/provisioning/ansible_common.html and ansible app task message below from playing on GCP stage environment (deleted after checking).
\- name: Show nginx_sites variable value
debug:
msg: "{{ nginx_sites }}"

Were installed Molecule, Ansible, Testinfra in python virtualenv environment. Changed testinfra version in requirements.txt because received an error:
_"molecule 2.14.0 has requirement testinfra==1.12.0, but you'll have testinfra 1.14.0 which is_ incompatible."
Were created and running some tests by Testinfra modules for check: MongoDB is enabled and running, has valid bindIp in config and listening TCP port 27017.
Packer templates was changed for use role.
For my WSL environment i needed configure Virtualbox provider across provider_raw_config_args variable in molecule.yml:
platforms:
\- name: instance
provider_raw_config_args:
\- "customize [ 'modifyvm', :id, '--uartmode1', 'disconnected' ]"
**
Moved DB role to other git repository and plugged it to "infra" repository. Checked this by creating terraform stage envinroment and playing site.yml - was succesful.
Added checking my DB role to TravisCI (used example):https://github.com/ivanov2103/ansible_role_dbserver - build status badge in title README.md.
Added build notification to my Slack channel #igor_ivanov: https://devops-team-otus.slack.com/archives/C9M5X748Y/p1529169830000075
In integration process I had some errors:
_"GoogleAuthError('PyCrypto library required for '\nlibcloud.common.google.GoogleAuthError: 'PyCrypto library required for Service Account Authentication.'\n"}
ERROR: Ansible version '2.3.0.0' not supported. Molecule only supports Ansible versions '>=2.5' with Python version '(3, 6)'_
For correcting this errors was changed install instruction in .travis.yml: "- **travis_wait** pip install **ansible>=2.5** molecule apache-libcloud **pycrypto**"

## Как запустить проект:
- commit changes DB role repository.
## Как проверить работоспособность:
- TravisCI build status must be passing.
## PR checklisk
- [*] Выставил label с номером домашнего задания
- [*] Выставил label с темой домашнего задания

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -36,7 +33,7 @@ script:
 - cd $MYHOME/terraform/stage/ ; mv backend.tf backend.tf.example && terraform init && terraform validate
 - cd $MYHOME/terraform/prod/ ; mv backend.tf backend.tf.example && terraform init && terraform validate
 - cd $MYHOME
-- ansible-lint ansible/playbooks/*
+- for file in `find ansible/ -type f -name "*.yml" ! -path "*jdauphant*"`; do ansible-lint $file; done
```
Comment (Nklya) [inline .travis.yml:36]: Аналогично, и в текущем виде ansible-lint не увидит папку с ролями и не заметит проблем в роли nginx

File: README.md [hunk 2]
```diff
@@ -101,3 +101,30 @@ Dynamic inventory from Homework-10 was configured for using in stage and prod en
...
+    provider_raw_config_args:  
+      \- "customize [ 'modifyvm', :id, '--uartmode1', 'disconnected' ]"  
+### **\*\***
+Moved DB role to other git repository and plugged it to "infra" repository. Checked this by creating terraform stage envinroment and playing site.yml - was succesful.  
+Added checking my DB role to TravisCI (used example):https://github.com/ivanov2103/ansible_role_dbserver - build status badge in title README.md.  
+Added build notification to my Slack channel \#igor_ivanov.  
+In integration process I had some errors:  
+*"GoogleAuthError('PyCrypto library required for '\nlibcloud.common.google.GoogleAuthError: 'PyCrypto library required for Service Account Authentication.'\n"}*  
+*ERROR: Ansible version '2.3.0.0' not supported.  Molecule only supports Ansible versions '>=2.5' with Python version '(3, 6)'*  
+For correcting this errors was changed install instruction in .travis.yml: "- **travis_wait** pip install **ansible>=2.5** molecule apache-libcloud **pycrypto**"
```
Comment (Nklya) [inline README.md:130]: 👍

**Review by Nklya — APPROVED** [review]
Добрый день.

Игорь, условия базового и дополнительного ДЗ выполнены, все реализовано хорошо 👍

File: .travis.yml (hunk 1 above)
Comment (ivanov2103) [reply .travis.yml:36]: При запуске из папки ansible, получаю ошибки, поищу, как решается.
ansible-lint ./*
Traceback (most recent call last):
  File "/usr/local/bin/ansible-lint", line 11, in <module>
    sys.exit(main())
  File "/usr/local/lib/python2.7/dist-packages/ansiblelint/__main__.py", line 173, in main
    matches.extend(runner.run())
  File "/usr/local/lib/python2.7/dist-packages/ansiblelint/__init__.py", line 229, in run
    for child in ansiblelint.utils.find_children(arg, self.playbook_dir):
  File "/usr/local/lib/python2.7/dist-packages/ansiblelint/utils.py", line 159, in find_children
    items = _playbook_items(playbook_ds)
  File "/usr/local/lib/python2.7/dist-packages/ansiblelint/utils.py", line 144, in _playbook_items
    return [item for play in pb_data for item in play.items()]
AttributeError: 'unicode' object has no attribute 'items'

Comment (Nklya) [reply .travis.yml:36]: Зачем ansible-lint ко всем файлам.
можно например только site.yml и пакеровские плейбуки. Или find.

Comment (ivanov2103) [reply .travis.yml:36]: Изначально так и было - только содержимое playbooks, где расположены site.yml и пакеровские плейбуки. Но на это было замечание и я сделал find. В jdauphant.nginx мы ничего не правим, используем, как есть, я ее исключил. Что тогда не так, поясни, пожалуйста.

Comment (Nklya) [reply .travis.yml:36]: В предыдущем PR было не так. ansible-lint выполнялся так: `ansible-lint ansible/playbooks/*`
https://github.com/Otus-DevOps-2018-02/ivanov2103_infra/pull/11/files#diff-354f30a63fb0907d4ad57269548329e3R39


Вот пример работающего решения. Здесь только site.yml, но можно добавить и пакеровские плейбуки.
https://github.com/express42/otus-homeworks/blob/2018-02/homeworks/ansible-3/controls/ansible.rb#L14

Comment (ivanov2103) [reply .travis.yml:36]: Ок, спасибо.

**Pull request merged**